src/services/anchor-service.ts

```diff
@@ -267,11 +256,31 @@ export default class AnchorService {
    * @param requests - array of anchor requests
    * @returns - map of docIds to 'Candidate' objects.
    */
-  async _groupCandidatesByDocId(requests: Request[]): Promise<Record<string, Candidate[]>> {
-    const groupedCandidates: Record<string, Candidate[]> = {};
+  async _groupCandidatesByDocId(requests: Request[]): Promise<Map<string, Candidate[]>> {
+    const groupedCandidates: Map<string, Candidate[]> = new Map();
+
+    let streamLimit = -1
+    if (config.merkleDepthLimit > 0) {
+      // The number of streams we are able to include in a single anchor batch is limited by the
+      // max depth of the merkle tree.
+      streamLimit = Math.pow(2, config.merkleDepthLimit)
+    }
 
     let request = null;
     for (let index = 0; index < requests.length; index++) {
+      if (streamLimit > 0 && groupedCandidates.size >= streamLimit) {
+        logger.warn('More than ' + groupedCandidates.size + ' candidate streams found, ' +
+          'which is the limit that can fit in a merkle tree of depth ' + config.merkleDepthLimit +
+          '. Returning unprocessed requests to PENDING status')
+
+        const unprocessedRequests = requests.slice(index)
+        await this.requestRepository.updateRequests({
```

Comment:
Not based on what we are currently tracking but will be good to keep in mind for the future

Reply:
So actually we don't want to include this `updateRequests`. We can just leave the status as processing because when we do `findNextToProcess` we expire the ones with processing status based on how long it's been since they were marked as such. With this change they would never get expired.

Reply:
That's exactly why we want this, we don't want these requests to get expired unnecessarily.  The expiration logic only applies to requests that have been in the PROCESSING state for a long time.  I assume that logic exists to catch requests that get stuck in PROCESSING somehow, without the processing ever actually finishing and either making the request COMPLETED or FAILED.  Leaving these requests in PROCESSING would make it look like they got stuck somehow, when in fact we just never actually even tried to process them in the first place

Reply:
Actually it matters less than I thought. The `expirationPeriod` just defines how long a request has to stay in PROCESSING before a future batch will attempt to re-process it again.  There's no actual proactive "expiration" where we actually fail the request because it "expired".

So the difference in behavior is that if put these requests back into PENDING, then we're guaranteed that the next batch considers them again.  If we leave them in PROCESSING, then they may not be considered in the next batch if the next batch happens faster than the `expirationPeriod`

Reply:
I don't know what the original intent was but if something passes the expiration period I would assume we are saying it's too late for that to get anchored. With your change we are saying, we will get to this eventually. Either is fine to me.

Reply:
Yep, I would assume that too from the word `expirationPeriod`, but reading the code that's not what happens.  When the expiration period passes the request is not failed. expirationPeriod shows up only once in the codebase, [here](https://github.com/ceramicnetwork/ceramic-anchor-service/blob/develop/src/repositories/request-repository.ts#L68-L78). That code is just selecting which Requests to consider processing in this batch.  So all `expirationPeriod` does is make us process requests that have been in the PROCESSING state for more than the configured time period.

So actually both with and without this code to put it back into PENDING the request will eventually get processed again, all that changes is how quickly it becomes eligible to processed in a subsequent batch

Reply:
Oh wow ok. I just realized that the ones left in processed get re-upped because we update everything each time.. This is quite confusing haha 😅 But yeah in this case put it back into pending feels like a bug fix in a way